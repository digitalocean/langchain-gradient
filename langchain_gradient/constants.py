# Constants for langchain_gradient

ALLOWED_MODEL_FIELDS = {
    "model_name",
    "frequency_penalty",
    "logit_bias",
    "logprobs",
    "max_completion_tokens",
    "max_tokens",
    "n",
    "presence_penalty",
    "stop",
    "streaming",
    "temperature",
    "top_logprobs",
    "top_p",
    "user",
    "timeout",
    "stream_options",
    "enable_rate_limiting",  # NEW
    "max_requests_per_minute",  # NEW
    # Add new fields here as needed
}
