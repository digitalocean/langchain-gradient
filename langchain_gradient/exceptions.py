class APITimeoutError(Exception):
    """Exception raised when API request times out."""
    
    def __init__(self, message: str = "API request timed out", timeout: float = None):
        self.timeout = timeout
        self.message = f"{message} (timeout={timeout}s)" if timeout else message
        super().__init__(self.message)


class APIRetryError(Exception):
    """Exception raised when max retries are exceeded."""
    
    def __init__(self, message: str = "Max retries exceeded", attempts: int = None):
        self.attempts = attempts
        self.message = f"{message} (attempts={attempts})" if attempts else message
        super().__init__(self.message)
