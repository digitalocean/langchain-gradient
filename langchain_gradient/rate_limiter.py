"""Rate limiting implementation for API requests."""

import time
from collections import deque
from threading import Lock


class RateLimiter:
    """
    Thread-safe rate limiter using token bucket algorithm.

    The token bucket algorithm allows for burst traffic while maintaining
    an average rate limit. Tokens are added at a fixed rate, and each
    request consumes one token.

    Attributes:
        max_requests: Maximum number of requests allowed in the time window
        time_window: Time window in seconds for the rate limit
        requests: Deque storing timestamps of recent requests
        lock: Threading lock for thread-safe operations

    Example:
        >>> limiter = RateLimiter(max_requests=60, time_window=60)
        >>> limiter.wait_if_needed()  # Waits if rate limit would be exceeded
    """

    def __init__(self, max_requests: int = 60, time_window: float = 60.0):
        """
        Initialize rate limiter.

        Args:
            max_requests: Maximum number of requests allowed in time window.
                         Default is 60 requests.
            time_window: Time window in seconds. Default is 60 seconds.

        Raises:
            ValueError: If max_requests or time_window are invalid.
        """
        if max_requests <= 0:
            raise ValueError(f"max_requests must be positive, got {max_requests}")
        if time_window <= 0:
            raise ValueError(f"time_window must be positive, got {time_window}")

        self.max_requests = max_requests
        self.time_window = time_window
        self.requests: deque = deque()
        self.lock = Lock()

    def wait_if_needed(self) -> None:
        """
        Wait if making a request now would exceed the rate limit.

        This method is thread-safe and can be called from multiple threads.
        It automatically removes expired request timestamps and calculates
        the required wait time if the rate limit would be exceeded.

        The method uses a sliding window approach: it only considers requests
        made within the current time window.
        """
        with self.lock:
            current_time = time.time()

            # Remove requests outside the current time window
            while self.requests and self.requests[0] <= current_time - self.time_window:
                self.requests.popleft()

            # Check if we need to wait
            if len(self.requests) >= self.max_requests:
                # Calculate wait time: time until oldest request expires
                oldest_request = self.requests[0]
                wait_time = self.time_window - (current_time - oldest_request)

                if wait_time > 0:
                    print(
                        f"⏳ Rate limit reached ({self.max_requests} requests "
                        f"per {self.time_window}s). Waiting {wait_time:.2f}s..."
                    )
                    time.sleep(wait_time)

                    # Remove the oldest request after waiting
                    self.requests.popleft()

            # Record this request
            self.requests.append(time.time())

    def get_current_usage(self) -> dict:
        """
        Get current rate limit usage statistics.

        Returns:
            Dictionary containing:
                - current_requests: Number of requests in current window
                - max_requests: Maximum allowed requests
                - time_window: Time window in seconds
                - usage_percentage: Percentage of rate limit used
        """
        with self.lock:
            current_time = time.time()

            # Remove expired requests
            while self.requests and self.requests[0] <= current_time - self.time_window:
                self.requests.popleft()

            current_requests = len(self.requests)
            usage_percentage = (current_requests / self.max_requests) * 100

            return {
                "current_requests": current_requests,
                "max_requests": self.max_requests,
                "time_window": self.time_window,
                "usage_percentage": round(usage_percentage, 2),
            }

    def reset(self) -> None:
        """
        Reset the rate limiter by clearing all request history.

        This is useful for testing or when you want to start fresh.
        """
        with self.lock:
            self.requests.clear()

    def __repr__(self) -> str:
        """String representation of the rate limiter."""
        usage = self.get_current_usage()
        return (
            f"RateLimiter("
            f"max_requests={self.max_requests}, "
            f"time_window={self.time_window}s, "
            f"current_usage={usage['current_requests']}/{self.max_requests})"
        )
