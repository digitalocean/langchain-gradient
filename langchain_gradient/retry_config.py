import time
import random
from typing import Callable, Optional, Tuple
from dataclasses import dataclass


@dataclass
class RetryConfig:
    """Configuration for retry behavior with exponential backoff."""
    
    max_retries: int = 3
    initial_delay: float = 1.0
    max_delay: float = 60.0
    exponential_base: float = 2.0
    jitter: bool = True
    retry_on_exceptions: Tuple = (Exception,)
    
    def calculate_delay(self, attempt: int) -> float:
        """Calculate delay for given attempt using exponential backoff."""
        delay = min(
            self.initial_delay * (self.exponential_base ** attempt),
            self.max_delay
        )
        
        if self.jitter:
            # Add random jitter (Â±20% of delay)
            jitter_range = delay * 0.2
            delay += random.uniform(-jitter_range, jitter_range)
        
        return max(0, delay)


def with_retry(
    retry_config: RetryConfig,
    timeout: Optional[float] = None
) -> Callable:
    """Decorator to add retry logic with exponential backoff."""
    
    def decorator(func: Callable) -> Callable:
        def wrapper(*args, **kwargs):
            last_exception = None
            
            for attempt in range(retry_config.max_retries + 1):
                try:
                    return func(*args, **kwargs)
                except retry_config.retry_on_exceptions as e:
                    last_exception = e
                    
                    if attempt < retry_config.max_retries:
                        delay = retry_config.calculate_delay(attempt)
                        time.sleep(delay)
                    else:
                        raise APIRetryError(
                            f"Failed after {retry_config.max_retries + 1} attempts",
                            attempts=retry_config.max_retries + 1
                        ) from last_exception
            
            return None
        
        return wrapper
    return decorator
