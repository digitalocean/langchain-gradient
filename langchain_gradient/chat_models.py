"""LangchainDigitalocean chat models."""
import os
import time
import random
from typing import Any, Dict, Iterator, List, Optional, Union
from gradient import Gradient
from gradient.exceptions import APITimeoutError, APIConnectionError, APIError
from langchain_core.callbacks import (
    CallbackManagerForLLMRun,
)
from langchain_core.language_models import BaseChatModel
from langchain_core.messages import (
    AIMessage,
    AIMessageChunk,
    BaseMessage,
)
from langchain_core.outputs import ChatGeneration, ChatGenerationChunk, ChatResult
from pydantic import Field, model_validator
from typing_extensions import TypedDict
from .constants import ALLOWED_MODEL_FIELDS


class StreamOptions(TypedDict, total=False):
    include_usage: bool


class ChatGradient(BaseChatModel):
    """
    ChatGradient model for DigitalOcean Gradient.
    
    This class provides an interface to the DigitalOcean Gradient chat completion API,
    compatible with LangChain's chat model interface. It supports both standard and streaming
    chat completions, and exposes model parameters for fine-tuning requests.
    
    Enhanced with configurable timeout and retry handling using exponential backoff.
    
    Parameters
    ----------
    api_key : str, optional
        DigitalOcean Gradient API key. If not provided, will use DIGITALOCEAN_INFERENCE_KEY
        environment variable.
    model_name : str
        Name of the model to use (e.g., "llama3.3-70b-instruct").
    timeout : float, optional
        Request timeout in seconds. Defaults to 60.0 seconds.
    max_retries : int, optional
        Maximum number of retry attempts for failed requests. Defaults to 3.
    stream_options : StreamOptions, optional
        Streaming configuration options.
    **kwargs : Any
        Additional model parameters from ALLOWED_MODEL_FIELDS.
    
    Examples
    --------
    Basic usage:
    >>> from langchain_gradient import ChatGradient
    >>> chat = ChatGradient(
    ...     api_key="your_api_key",
    ...     model_name="llama3.3-70b-instruct",
    ...     timeout=30.0,
    ...     max_retries=5
    ... )
    
    With custom timeout configuration:
    >>> import httpx
    >>> chat = ChatGradient(
    ...     api_key="your_api_key", 
    ...     model_name="llama3.3-70b-instruct",
    ...     timeout=httpx.Timeout(60.0, read=5.0, write=10.0, connect=2.0)
    ... )
    """
    
    api_key: Optional[str] = Field(default=None, description="DigitalOcean Gradient API key")
    model_name: str = Field(description="Model name to use")
    timeout: Union[float, Any] = Field(default=60.0, description="Request timeout in seconds")
    max_retries: int = Field(default=3, description="Maximum retry attempts")
    stream_options: Optional[StreamOptions] = Field(default=None)
    
    # Add all allowed model fields as optional parameters
    temperature: Optional[float] = Field(default=None)
    max_tokens: Optional[int] = Field(default=None)
    top_p: Optional[float] = Field(default=None)
    frequency_penalty: Optional[float] = Field(default=None)
    presence_penalty: Optional[float] = Field(default=None)
    stop: Optional[Union[str, List[str]]] = Field(default=None)
    
    _client: Optional[Gradient] = None
    
    class Config:
        """Configuration for this pydantic object."""
        extra = "forbid"
        arbitrary_types_allowed = True
    
    @model_validator(mode="before")
    @classmethod
    def validate_environment(cls, values: Dict[str, Any]) -> Dict[str, Any]:
        """Validate that api key exists in environment."""
        api_key = values.get("api_key") or os.getenv("DIGITALOCEAN_INFERENCE_KEY")
        if not api_key:
            raise ValueError(
                "DigitalOcean Gradient API key not found. "
                "Please provide api_key parameter or set DIGITALOCEAN_INFERENCE_KEY environment variable."
            )
        values["api_key"] = api_key
        return values
    
    @property
    def client(self) -> Gradient:
        """Get or create the Gradient client with timeout and retry configuration."""
        if self._client is None:
            self._client = Gradient(
                inference_key=self.api_key,
                timeout=self.timeout,
                max_retries=self.max_retries
            )
        return self._client
    
    @property
    def _llm_type(self) -> str:
        """Return type of language model used by this chat model."""
        return "gradient"
    
    def _get_model_params(self) -> Dict[str, Any]:
        """Get model parameters for API calls."""
        params = {"model": self.model_name}
        
        # Add optional parameters if they are set
        for field in ALLOWED_MODEL_FIELDS:
            value = getattr(self, field, None)
            if value is not None:
                params[field] = value
        
        if self.stream_options:
            params["stream_options"] = self.stream_options
        
        return params
    
    def _format_messages(self, messages: List[BaseMessage]) -> List[Dict[str, str]]:
        """Format LangChain messages for Gradient API."""
        formatted_messages = []
        for message in messages:
            if hasattr(message, 'role'):
                role = message.role
            elif hasattr(message, 'type'):
                # Map LangChain message types to chat roles
                role_mapping = {
                    'human': 'user',
                    'ai': 'assistant', 
                    'system': 'system'
                }
                role = role_mapping.get(message.type, 'user')
            else:
                role = 'user'
            
            formatted_messages.append({
                "role": role,
                "content": message.content
            })
        return formatted_messages
    
    def _execute_with_retry(self, func, *args, **kwargs) -> Any:
        """Execute a function with exponential backoff retry logic."""
        last_exception = None
        
        for attempt in range(self.max_retries + 1):
            try:
                return func(*args, **kwargs)
            except APITimeoutError as e:
                last_exception = e
                if attempt == self.max_retries:
                    raise APITimeoutError(f"Request timed out after {self.max_retries} retries: {str(e)}") from e
                
                # Exponential backoff: 2^attempt + random jitter
                delay = min(60, (2 ** attempt) + random.uniform(0, 1))
                time.sleep(delay)
                
            except (APIConnectionError, APIError) as e:
                last_exception = e
                # Check if this is a retryable error
                if self._is_retryable_error(e):
                    if attempt == self.max_retries:
                        raise type(e)(f"Request failed after {self.max_retries} retries: {str(e)}") from e
                    
                    # Exponential backoff
                    delay = min(60, (2 ** attempt) + random.uniform(0, 1))
                    time.sleep(delay)
                else:
                    # Non-retryable error, raise immediately
                    raise
            
            except Exception as e:
                # For other exceptions, don't retry
                raise
        
        # This should never be reached, but just in case
        if last_exception:
            raise last_exception
    
    def _is_retryable_error(self, error: Exception) -> bool:
        """Determine if an error is retryable based on Gradient SDK patterns."""
        if isinstance(error, APITimeoutError):
            return True
        
        if isinstance(error, APIConnectionError):
            return True
            
        if hasattr(error, 'status_code'):
            # Retry on 408 Request Timeout, 409 Conflict, 429 Rate Limit, >=500 Internal errors
            retryable_codes = {408, 409, 429} | set(range(500, 600))
            return error.status_code in retryable_codes
        
        return False
    
    def _generate(
        self,
        messages: List[BaseMessage],
        stop: Optional[List[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> ChatResult:
        """Generate a chat completion response."""
        formatted_messages = self._format_messages(messages)
        params = self._get_model_params()
        
        if stop:
            params["stop"] = stop
        
        # Override with any additional kwargs
        params.update(kwargs)
        
        def _make_request():
            return self.client.chat.completions.create(
                messages=formatted_messages,
                **params
            )
        
        try:
            completion = self._execute_with_retry(_make_request)
            
            # Extract the response
            if completion.choices:
                message_content = completion.choices[0].message.content
                ai_message = AIMessage(content=message_content)
                generation = ChatGeneration(message=ai_message)
                
                # Add usage metadata if available
                llm_output = {}
                if hasattr(completion, 'usage') and completion.usage:
                    llm_output["usage"] = {
                        "prompt_tokens": getattr(completion.usage, "prompt_tokens", None),
                        "completion_tokens": getattr(completion.usage, "completion_tokens", None),
                        "total_tokens": getattr(completion.usage, "total_tokens", None),
                    }
                
                return ChatResult(generations=[generation], llm_output=llm_output)
            else:
                # No choices returned
                ai_message = AIMessage(content="")
                generation = ChatGeneration(message=ai_message)
                return ChatResult(generations=[generation])
                
        except Exception as e:
            # Handle any remaining errors
            error_message = f"Gradient API error: {str(e)}"
            ai_message = AIMessage(content=f"[ERROR] {error_message}")
            generation = ChatGeneration(message=ai_message)
            return ChatResult(generations=[generation])
    
    def _stream(
        self,
        messages: List[BaseMessage],
        stop: Optional[List[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> Iterator[ChatGenerationChunk]:
        """Stream chat completion responses."""
        formatted_messages = self._format_messages(messages)
        params = self._get_model_params()
        params["stream"] = True
        
        if stop:
            params["stop"] = stop
        
        # Override with any additional kwargs
        params.update(kwargs)
        
        def _make_stream_request():
            return self.client.chat.completions.create(
                messages=formatted_messages,
                **params
            )
        
        try:
            stream = self._execute_with_retry(_make_stream_request)
            
            for chunk in stream:
                if chunk.choices:
                    choice = chunk.choices[0]
                    if hasattr(choice, 'delta') and choice.delta:
                        content = getattr(choice.delta, 'content', '') or ''
                        if content:
                            ai_chunk = AIMessageChunk(content=content)
                            yield ChatGenerationChunk(message=ai_chunk)
            
            # Optionally yield usage metadata at the end if available
            if self.stream_options and self.stream_options.get("include_usage"):
                # This would be handled by the stream completion itself in real usage
                pass
                        
        except Exception as e:
            # Yield an error chunk if possible
            error_chunk = ChatGenerationChunk(
                message=AIMessageChunk(
                    content=f"[ERROR] {str(e)}", 
                    response_metadata={"error": str(e)}
                )
            )
            yield error_chunk
    
    @property
    def init_from_env_params(self) -> tuple[dict, dict, dict]:
        # env_vars, model_params, expected_attrs
        # Map DIGITALOCEAN_INFERENCE_KEY -> api_key, and require model param
        return (
            {"DIGITALOCEAN_INFERENCE_KEY": "test-env-key"},
            {"model": "bird-brain-001", "buffer_length": 50},
            {"api_key": "test-env-key", "model_name": "bird-brain-001"},
        )
    
    @classmethod
    def is_lc_serializable(cls) -> bool:
        return True
    
    def __getstate__(self) -> dict:
        state = self.__dict__.copy()
        # Exclude sensitive credentials from serialization
        state.pop("api_key", None)
        return state
    
    def __setstate__(self, state: dict) -> None:
        self.__dict__.update(state)
