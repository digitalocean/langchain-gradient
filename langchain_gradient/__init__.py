"""LangChain integration with DigitalOcean Gradient."""

from langchain_gradient.chat_models import (
    ChatGradient,
    StructuredOutputError,
    StructuredOutputRunnable,
)

__all__ = [
    "ChatGradient",
    "StructuredOutputError", 
    "StructuredOutputRunnable",
]
