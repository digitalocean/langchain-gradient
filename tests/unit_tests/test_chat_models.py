"""Test chat model integration."""

import os
from typing import Type

from dotenv import load_dotenv
from langchain_tests.unit_tests import ChatModelUnitTests

from langchain_gradient.chat_models import ChatGradient

load_dotenv()


class TestChatGradientUnit(ChatModelUnitTests):
    @property
    def chat_model_class(self) -> Type[ChatGradient]:
        return ChatGradient

    @property
    def chat_model_params(self) -> dict:
        # These should be parameters used to initialize your integration for testing
        return {
            "model": "llama3.3-70b-instruct",
            "temperature": 0,
            "api_key": os.environ.get("DIGITALOCEAN_INFERENCE_KEY"),
        }
