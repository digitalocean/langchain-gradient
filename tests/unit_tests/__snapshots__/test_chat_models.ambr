# serializer version: 1
# name: TestChatGradientUnit.test_serdes[serialized]
  dict({
    'id': list([
      'langchain_gradient',
      'chat_models',
      'ChatGradient',
    ]),
    'kwargs': dict({
      'max_retries': 2,
      'max_tokens': 100,
      'model_name': 'llama3.3-70b-instruct',
      'temperature': 0.0,
      'timeout': 60.0,
      'user': 'langchain-gradient',
    }),
    'lc': 1,
    'name': 'ChatGradient',
    'type': 'constructor',
  })
# ---
